[MRG] Allow only fit_transform to be present in pipeline

jnboehm opened the pull request:
This makes it possible to plug in algorithms like TSNE in the
pipeline, although they do not support `transform`.  A warning will be
raised that the pipeline is not reusable.

#### Reference Issues/PRs
Fixes  #16710

#### What does this implement/fix? Explain your changes.
This allows algorithms that only implement a `fit_transform` function, but not `transform` (such as TSNE) to be used in a Pipeline in sklearn.

Commit:
Allow only fit_transform to be present in pipeline

This makes it possible to plug in algorithms like TSNE in the
pipeline, although they do not support `transform`.  A warnign will be
raised that the pipeline is not reusable.

Commit:
Fix linter error

johannfaouzi commented on sklearn/pipeline.py:
```diff
@@ -177,19 +177,25 @@ def _validate_steps(self):
-            if (not (hasattr(t, "fit") or hasattr(t, "fit_transform")) or not
-                    hasattr(t, "transform")):
+            if (not hasattr(t, "fit_transform") or
+                    (not hasattr(t, "fit") and hasattr(t, "transform"))):
```
It should be 
```python
(not (hasattr(t, "fit") and hasattr(t, "transform")))):
```
no? I think that you need parentheses to say `not (fit and transform)`.

jnboehm replied:
Yes, that's correct, not has precedence over and. I'll fix it

johannfaouzi commented on sklearn/pipeline.py:
```diff
@@ -177,19 +177,25 @@ def _validate_steps(self):
+                                "transform, fit_transform or be the string "
+                                "'passthrough'. '%s' (type %s) doesn't"
+                                % (t, type(t)))
+            elif not hasattr(t, "transform"):
```
Don't you want to also raise a warning if the last step doesn't have a `transform` or `predict` method? The pipeline is also not reusable in those cases.

jnboehm replied:
Right, I added a a check for the final estimator as well.

johannfaouzi reviewed (commented):
Thanks! It would be nice if you could add some tests in the `sklearn/tests/test_pipeline.py` file to make sure that the warning and the error are raised in relevant cases.

Commit:
Check final estimator

Commit:
Fix faultly if condition

johannfaouzi commented on sklearn/pipeline.py:
```diff
@@ -177,20 +177,31 @@ def _validate_steps(self):
 
         # We allow last estimator to be None as an identity transformation
         if (estimator is not None and estimator != 'passthrough'
-                and not hasattr(estimator, "fit")):
```
You need to remove the `:` at the end of this line.

johannfaouzi commented on sklearn/pipeline.py:
```diff
@@ -177,20 +177,31 @@ def _validate_steps(self):
+                  not hasattr(estimator, "predict")):
+            warnings.warn("Intermediate step '%s' (type %s) does not have "
+                          "transform or predict, pipeline is not reusable."
+                          % (estimato, type(estimator)))
```
`estimator` instead of `estimato` ;)

jnboehm replied:
Oh God, yes.  Thank you for catching this.

Commit:
Fix variable typo

Commit:
Add a test case for the new pipeline construction

johannfaouzi commented on sklearn/tests/test_pipeline.py:
```diff
@@ -392,6 +393,22 @@ def test_pipeline_methods_preprocessing_svm():
 
         pipe.score(X, y)
 
+def test_pipeline_methods_pca_tsne():
```
You need one more blank line above (two blank lines instead of one).

jnboehm replied:
Pushed a fix

johannfaouzi commented on sklearn/pipeline.py:
```diff
@@ -177,20 +177,31 @@ def _validate_steps(self):
+                "fit_transform or be the string 'passthrough'. "
                 "'%s' (type %s) doesn't" % (estimator, type(estimator)))
+        elif (not hasattr(estimator, "transform") and
+                  not hasattr(estimator, "predict")):
```
You can split this into 4 lines (because of PEP8):
```python
elif (
    not hasattr(estimator, "transform") and
    not hasattr(estimator, "predict")
):
```
or rephrase it (putting `not` before the parenthesis and replacing `and` with `or`):
```python
elif not (hasattr(estimator, "transform") or
          hasattr(estimator, "predict")):

```

jnboehm replied:
The latter one looks nicer to me, will change it to that

Commit:
Expand test cases with warning and error

jnboehm commented:
> Thanks! It would be nice if you could add some tests in the `sklearn/tests/test_pipeline.py` file to make sure that the warning and the error are raised in relevant cases.

Sure! Didn't see that earlier.

Commit:
Fix pep8 style issue

According to
https://github.com/scikit-learn/scikit-learn/pull/16714#discussion_r393853164

changed the syntax but not semantics

Commit:
Add blank lines

According to the style guide

johannfaouzi commented on sklearn/tests/test_pipeline.py:
```diff
@@ -393,6 +394,36 @@ def test_pipeline_methods_preprocessing_svm():
+    separate_emb = tsne.fit_transform(pca.fit_transform(iris.data))
+
+    pca_for_pipline = PCA(n_components=2, random_state=0)
+    tsne_for_pipline = TSNE(random_state=0)
```
missing `e` (`pipline` instead of `pipeline`)

Commit:
Fix variable names in typos

johannfaouzi commented on sklearn/pipeline.py:
```diff
@@ -177,20 +177,31 @@ def _validate_steps(self):
-            if (not (hasattr(t, "fit") or hasattr(t, "fit_transform")) or not
-                    hasattr(t, "transform")):
+            if (not hasattr(t, "fit_transform") or
+                    (not (hasattr(t, "fit") and hasattr(t, "transform")))):
```
It should be an `and` at the end of line 180, no?
```python
if not fit_transform AND not(fit and transform)
```
If it doesn't have a `fit_transform` method, the condition will be True currently. That's why some tests are failing.

jnboehm replied:
Yeah, that makes sense.

jnboehm replied:
Also fixed it for `FeatureUnion._validate_transformers()`.

Commit:
Fix conditional for pipeline error check

Commit:
Fix conditional for feature union error check

johannfaouzi commented on sklearn/tests/test_pipeline.py:
```diff
@@ -393,6 +394,36 @@ def test_pipeline_methods_preprocessing_svm():
+           "transform, pipeline is not reusable." % (tsne_for_pipeline,
+                                                     type(tsne_for_pipeline)))
+    with pytest.warns(UserWarning, match=msg):
+        pipe = make_pipeline(pca_for_pipeline, tsne_for_pipeline)
```
The warning is not raised

jnboehm replied:
This has been fixed now, it was an issue with the regex match, that should be fixed with explicit strings and callng to `re.escape`.

johannfaouzi reviewed (commented):
* You have to update the error message in this [test](https://github.com/jnboehm/scikit-learn/blob/2a295cf769c761869ec0585d6d177e17f9c31ba9/sklearn/tests/test_pipeline.py#L168-L170).

* You have to update the error message in this [test](https://github.com/jnboehm/scikit-learn/blob/2a295cf769c761869ec0585d6d177e17f9c31ba9/sklearn/tests/test_pipeline.py#L525-L526).

Commit:
Fix raised error messages

Also change the test for FeatureUnion

Commit:
Fix linting

jnothman commented on sklearn/pipeline.py:
```diff
@@ -177,20 +177,31 @@ def _validate_steps(self):
+                                % (t, type(t)))
+            elif not hasattr(t, "transform"):
+                warnings.warn("Intermediate step '%s' (type %s) does not have "
+                              "transform, pipeline is not reusable."
```
maybe "on test data"

jnboehm replied:
Good point, I reworded the warning

jnothman commented on sklearn/pipeline.py:
```diff
@@ -177,20 +177,31 @@ def _validate_steps(self):
+                "Last step of Pipeline should implement fit, "
+                "fit_transform or be the string 'passthrough'. "
                 "'%s' (type %s) doesn't" % (estimator, type(estimator)))
+        elif not (hasattr(estimator, "transform") or
```
This is not an intermediate step. We do not need to handle the case you've handled here.

johannfaouzi replied:
Don't you think it is worth raising a warning in that case? Or the user should expect this since it's the final step of the pipeline, and the existing code (that checks which methods are available for the last step) is enough?

jnboehm replied:
Personally, I would find it more elegant to remove that code and leave it up to the user.  If somebody throws something like `TSNE` in there, I would expect that person to know what to do.

jnothman commented on sklearn/pipeline.py:
```diff
@@ -177,20 +177,31 @@ def _validate_steps(self):
         if (estimator is not None and estimator != 'passthrough'
-                and not hasattr(estimator, "fit")):
+                and not (hasattr(estimator, "fit") or
+                         hasattr(estimator, "fit_transform"))):
```
I don't think we have or allow estimators that support fit_transform but not fit.

jnboehm replied:
Okay, I removed that check

jnothman commented on sklearn/pipeline.py:
```diff
@@ -864,11 +875,15 @@ def _validate_transformers(self):
                 continue
             if t == 'drop':
                 continue
-            if (not (hasattr(t, "fit") or hasattr(t, "fit_transform")) or not
```
I don't see any need for change here. there's no such thin as an "intermediate step" in a FeatureUnion

jnboehm replied:
Alright, I changed that code back along with the rewritten test

jnothman reviewed (commented):
Thanks

Commit:
Improve warning for reusable pipe warning

from comment
https://github.com/scikit-learn/scikit-learn/pull/16714#discussion_r394337647

Commit:
Remove last check in pipeline for final estimator

Should be taken care of by the user

Commit:
Remove unnecessary check for fit_transform attr

As per
https://github.com/scikit-learn/scikit-learn/pull/16714#discussion_r394338624

Commit:
Fix over-indented line

flake8 E127

Commit:
Revert changes to FeatureUnion

Commit:
Fix erroneous reverting reverting

Commit:
Modify warning message

Commit:
Remove unnecessary print params

johannfaouzi commented on sklearn/pipeline.py:
```diff
@@ -867,8 +872,11 @@ def _validate_transformers(self):
-                                "transform. '%s' (type %s) doesn't" %
-                                (t, type(t)))
+                                "transform. '%s' (type %s) doesn't"
+                                % (t, type(t)))
```
You can put the `%` on the previous line so that it does not change the existing code.

jnboehm replied:
That would amount to more than 79 characters in that line, wouldn't it?

johannfaouzi replied:
Sorry, it's not something that can easily be seen on GitHub only. Alright then!

jnboehm replied:
Oh, I misunderstood you, you're right. I changed it back to the previous version

johannfaouzi reviewed (commented):
There seems to be an issue with some indentation at line 874 but I'm not sure why:
https://dev.azure.com/scikit-learn/scikit-learn/_build/results?buildId=14629&view=logs&j=9219275c-2b6e-53e9-501f-47c0b597ffd4&t=3c16a9c6-c72d-50f1-8590-5824faaa9233&l=1298

Maybe a mix of tabs and 4 whitespaces?

Commit:
Match literal strings in pytest.warns and raises

This should take care of the uncaught warnings that have been emitted before

Commit:
Pass the estimator to the pipeline

Otherwise it can hardly raise the error

Commit:
Add dummy transformer to trigger correct warning

Commit:
Move the correct code line into the pytest context

The error is raised during creation of the pipeline, not when calling
fit_transform.

Remove dummy transformer

Commit:
Fix error raising for degenerate pipeline

The final step does not necessarily need to implement a transform,
hence the warning aws rightfully not raised before.

jnboehm commented:
Alright, I have finally managed to wrap my head around `pytest` and manage to complete the test suite.  Let me know whether anything should be done or corrected before merging the PR.  Sorry forthe noisy commits and tahnks for your patience (thus far).

johannfaouzi commented on sklearn/tests/test_pipeline.py:
```diff
@@ -393,6 +394,39 @@ def test_pipeline_methods_preprocessing_svm():
+        make_pipeline(error_estimator, 'passthrough')
+
+
+
```
Only 2 blank lines (instead of 3).

johannfaouzi reviewed (approved):
One minor comment otherwise LGTM! Thanks @jnboehm!

You can let maintainers know that you are done by adding `[MRG]` at the beginning of the title of your PR.

Commit:
Remove a blank line

Commit:
Change unnecessary code format back to original

jnothman commented on sklearn/pipeline.py:
```diff
@@ -869,6 +874,9 @@ def _validate_transformers(self):
                                 (t, type(t)))
+            elif not hasattr(t, "transform"):
+                warnings.warn("Intermediate step '%s' (type %s) does not have "
+                              "transform" % (t, type(t)))
```
Maybe add "This FeatureUnion will not be usable with test data"

jnothman commented on sklearn/pipeline.py:
```diff
@@ -869,6 +874,9 @@ def _validate_transformers(self):
                                 "transform. '%s' (type %s) doesn't" %
                                 (t, type(t)))
+            elif not hasattr(t, "transform"):
+                warnings.warn("Intermediate step '%s' (type %s) does not have "
```
Intermediate step -> transformer

This change isn't tested. 

I'm also happy to see it removed

jnboehm replied:
Do you mean removing the entire warning altogether?  I thought that `FeatureUnion` should be similar in behaviour to Pipeline.

jnothman replied:
It probably should behave similar to Pipeline, but it's somewhat less valuable where `transform` is not supported! I'd also be happy to pragmatically merge functionality for Pipeline and deal with FeatureUnion later in a separate pull request.

jnboehm replied:
Alright, I have removed it for now.

jnothman commented on sklearn/tests/test_pipeline.py:
```diff
@@ -393,6 +394,38 @@ def test_pipeline_methods_preprocessing_svm():
+           "transform, pipeline is not reusable on test data."
+           % (tsne_for_pipeline, type(tsne_for_pipeline)))
+
+    with pytest.warns(UserWarning, match=re.escape(msg)):
```
In most estimators, validation is delayed until fit is called. While this isn't the case for Pipeline, it would be best to check for this warning only when fit_transform is called

jnboehm replied:
Do you mean that the check should be delayed until `fit_transform` is called or that the function should be called in the `pytest.warns` scope?

jnothman replied:
I think I mean that the check should be delayed until `fit_transform` is called. But I'm not entirely sure how the latter proposal differs.

jnboehm replied:
I initially thought that the call to `self._validate_steps()` should have been removed from `Pipeline.__init__()`.  I've changed the context manager now to guard `fit_transform`

jnothman commented on sklearn/tests/test_pipeline.py:
```diff
@@ -393,6 +394,38 @@ def test_pipeline_methods_preprocessing_svm():
+        pipe = make_pipeline(pca_for_pipeline, tsne_for_pipeline,
+                             'passthrough')
+
+    pipeline_emb = pipe.fit_transform(iris.data)
```
We should also test that fit_predict works if the final estimator is a clusterer

jnothman reviewed (commented):
Thanks

Please add a `|Fix|` entry to the change log at `doc/whats_new/v0.23.rst`. Like the other entries there, please reference this pull request with `:pr:` and credit yourself (and other contributors if applicable) with `:user:`

jnboehm committed:
Expand FeatureUnion warning

jnboehm committed:
Fix typo

jnboehm committed:
Add entry to the changelog

jnboehm committed:
Remove change in FeatureUnion

As per https://github.com/scikit-learn/scikit-learn/pull/16714#discussion_r410919097

jnboehm committed:
Delay warning test until fit_transform is called

https://github.com/scikit-learn/scikit-learn/pull/16714#discussion_r410919270

jnboehm committed:
Add test for nonreusable pipe with fit_predict

As suggested in
https://github.com/scikit-learn/scikit-learn/pull/16714#discussion_r401191598

jnboehm committed:
Fix linting error

jnboehm committed:
Accept y param for dummy fit_transform class

jnboehm commented:
I'm a bit unsure why the doc test failed, since I didn't change anything there. Could it be related to https://github.com/scikit-learn/scikit-learn/issues/17051?

thomasjpfan commented on doc/whats_new/v0.23.rst:
```diff
@@ -411,6 +411,14 @@ Changelog
   :pr:`16257` by :user:`Simona Maggio <simonamaggio>`.
 
+:mod:`sklearn.pipeline`
+......................
```
```suggestion
.......................
```

thomasjpfan reviewed (commented):
Merging with master may fix your issue.

jnboehm committed:
Add missing point

Update doc/whats_new/v0.23.rst

Co-Authored-By: Thomas J Fan <thomasjpfan@gmail.com>

jnboehm committed:
Merge remote-tracking branch 'origin/master' into pipeline/fit_transform

jnboehm committed:
Remove unncessary call to fit()

jnboehm committed:
Fix type in test case for warning

jnboehm committed:
Empty commit to restart CI pipeline

jnboehm commented:
The CI crashed again, but now everything seems to be working.  Let me know if anything needs fixing.

jnboehm committed:
Merge branch 'master' into pipeline/fit_transform

lucyleeow commented:
ping @cmarmo this might be worth a 'waiting for reviewer' tag

jnothman reviewed (commented):
So my only concern here is someone training a pipeline with a tsne in the middle, expecting that they'll then be able to predict with it. They wait an hour for the model to fit and then ... bam! No predictions for you. I don't want to waste a user's time like that.

The only way I can see around that is to have a parameter control whether this feature is available, or going further, to introduce a TransductivePipeline.

Pipeline is very much designed to solve needs around train/test setups, so I don't want to weaken its usability for that use-case.

Your thoughts?

lorentzenchr commented:
Shall we open a dedicated issue to discuss such usage patterns? It also popped up in #22162.

I agree with @jnothman 
> Pipeline is very much designed to solve needs around train/test setups, so I don't want to weaken its usability for that use-case.

thomasjpfan commented:
> Shall we open a dedicated issue to discuss such usage patterns? It also popped up in #22162.

Yes.
